docs/spec/BeakerSpec.tex

```diff
@@ -0,0 +1,994 @@
+index 0x02) held by this procedure.
+
+The very last value is an offset into the capability data. The meaning of this
+is different for each capability as they each have different formats. The format
+of each capability is specified in Section~\ref{syscalls-and-caps}
+
+\includegraphics[width=\textwidth]{media/ProcedureKeyLocation.pdf}
+
+The following table outlines what data is found at each of the different
+locations specified by the \emph{Data Offset}.
+
+\begin{table}[H]
+  \caption{Procedure data.}
+  \centering{}%
+  % \begin{tabular}{>{\centering}p{0.25\textwidth}|>{\centering}p{0.475\textwidth}|>{\centering}p{0.2\textwidth}}
+  \begin{tabularx}{\textwidth}{c | c | c | c | X }
+    \hline
+    \multicolumn{4}{c|}{Data Offset} & Description \\
+    \hline
+    \hline
+    \texttt{0x} & \texttt{00} & \texttt{00} & \texttt{00} & address: 20 bytes,
+      aligned right in the 32 bytes \\
+    \texttt{0x} & \texttt{$ty$} & \texttt{00} & \texttt{00} & The number of
+      capabilities of type \texttt{0x$ty$}. \\
+    \texttt{0x} & \texttt{$ty$} & \texttt{$in$} & \texttt{$of$} & Capability of
+      type \texttt{0x$ty$}, with index \texttt{0x$in-1$}, and offset
+      \texttt{0x$of$} into that capability. \\
+    \hline
+  \end{tabularx}
+\end{table}
+
+\section{System Calls and Capabilities}\label{syscalls-and-caps}
+All system calls are transactions to the kernel. The transaction data is defined
+as follows:
+
+\begin{table}[H]
+  \caption{System call structure.}
+  \centering{}%
+  \begin{tabular}{l|p{0.7\textwidth}}
+    \hline
+    Byte Offset & Description\tabularnewline
+    \hline
+    \hline
+    0x0 & 1 byte, system call value \tabularnewline
+    0x8 & The system call data as defined by each system call \tabularnewline
+    ~~~$\vdots$ &  \tabularnewline
+    \hline
+  \end{tabular}
+\end{table}
+
+If the transaction data is longer that specified by the system call format, the
+additional data is simply ignored. \textbf{TODO:} We should consider throwing an
+error on this.
+
+System call numbers match capability numbers.
+
+\begin{table}[H]
+  \caption{System call and capability numbers.\label{syscall-numbers}}
+  \centering{}%
+  \begin{tabular}{l|p{0.7\textwidth}}
+    \hline
+    Type Value & Description\tabularnewline
+    \hline
+    \hline
+    0x0 & Null capability / noop \tabularnewline
+    % 0x1 & The system call data as defined by each system call \tabularnewline
+    % 0x2 & The system call data as defined by each system call \tabularnewline
+    0x3 & Call Procedure \tabularnewline
+    0x4 & Register Procedure \tabularnewline
+    0x5 & Delete Procedure \tabularnewline
+    0x6 & Set Entry Procedure \tabularnewline
+    0x7 & Write \tabularnewline
+    0x8 & Log \tabularnewline
+    0x9 & Gas Send \tabularnewline
+    \hline
+  \end{tabular}
+\end{table}
+
+\subsection{Executing a System Call}
+The running code will issue a transaction to the kernel by perfoming a
```

Comment:
So right now the kernel instance contains both the kernel logic and storage, which will have to change unless we decide to keep the two together.

Reply:
We also need to account for procedure selfdestruct attacks which we haven not discussed #115

Reply:
This should be covered by the execution guard now.